lumeer-core/src/main/java/io/lumeer/core/facade/LinkInstanceFacade.java

```diff
@@ -257,6 +257,29 @@ public LinkInstance getLinkInstance(String linkTypeId, String linkInstanceId) {
       return stored;
    }
 
+   public Set<LinkInstance> duplicateLinkInstances(final String originalDocumentId, final String newDocumentId, final Set<String> linkInstanceIds) {
+      final Set<LinkInstance> result = new HashSet<>();
+      final Map<String, LinkType> allowedLinkTypeIds = new HashMap<>();
+
+      linkInstanceIds.forEach(id -> {
+         final LinkInstance link = linkInstanceDao.getLinkInstance(id);
+         final LinkType linkType = allowedLinkTypeIds.computeIfAbsent(link.getLinkTypeId(), this::checkLinkTypeWritePermissions);
+
+         final LinkInstance newLink = linkInstanceDao.duplicateLinkInstance(link, originalDocumentId, newDocumentId);
+
+         if (newLink != null) {
+            final DataDocument originalData = linkDataDao.getData(link.getLinkTypeId(), id);
+            final DataDocument newData = linkDataDao.createData(link.getLinkTypeId(), newLink.getId(), originalData);
+            constraintManager.decodeDataTypes(linkType, newData);
+            newLink.setData(newData);
+
+            result.add(newLink);
+         }
+      });
```

Comment:
Imagine situation, when document has i.e. 1000 documents. This code call database 4000 times :( 
I know we don't have solution for problems like this, but I'm afraid we'll have big performance problems soon.

Reply:
There is no other possibility, I cannot load all the data into memory. I could load all link instances, but definitely not data... So WDYT?

Reply:
Is it better now?